src/kernel/arch/x86/syscalls.zig

```
@@ -0,0 +1,173 @@
+    return syscall < NUM_HANDLERS;
+}
+
+///
+/// Handle a syscall. Gets the syscall number from eax within the context and calls the registered handler.
+/// If there isn't a registered handler or the syscall is invalid (>= NUM_HANDLERS) then a warning is logged
+///
+/// Arguments:
+///     IN ctx: *arch.InterruptContext - The cpu context when the syscall was triggered. The syscall number is
+///                                     stored in eax.
+///
+fn handle(ctx: *arch.InterruptContext) void {
+    // The syscall number is put in eax
+    const syscall = ctx.eax;
+    if (isValidSyscall(syscall)) {
+        if (handlers[syscall]) |handler| {
+            handler(ctx);
+        } else {
+            log.logWarning("Syscall {} triggered but not registered\n", syscall);
+        }
+    } else {
+        log.logWarning("Syscall {} is invalid\n", syscall);
+    }
+}
+
+///
+/// Register a syscall so it can be called by triggering interrupt 128
+///
+/// Arguments:
+///     IN syscall: u8 - The syscall number to register the handler with
+///     IN handler: SyscallHandler - The handler to register the syscall with
+///
+/// Errors:
+///     SyscallError.InvalidSyscall - If the syscall is invalid (see isValidSyscall)
+///     SyscallError.SyscallExists - If the syscall has already been registered
+///
+pub fn registerSyscall(syscall: u8, handler: SyscallHandler) !void {
+    if (!isValidSyscall(syscall))
+        return SyscallError.InvalidSyscall;
+    if (handlers[syscall]) |_|
+        return SyscallError.SyscallExists;
+    handlers[syscall] = handler;
+}
+
+///
+/// Trigger a syscall. Returns the value put in eax by the syscall handler.
+///
+/// Arguments:
+///     IN syscall: u8 - The syscall to trigger. Is put in eax and must have been registered.
+///     IN args: u32... - The args to send to the syscall. A maximum of 5 are put in ebx, ecx, edx, esi and edi
+///
+pub inline fn doSyscall(comptime syscall: u32, args: ...) u32 {
```

Also, is it worth adding a check for valid syscalls or is it fine to use the check after the syscall

---

Hmm good question, reckon we should return an error if it's an invalid syscall?

---

Yeh, maybe the same error for the kernel one? Or a separate error